File: packages/agora-agent-sdk-go/protocol.go
// Package agorasdk provides a Go client for the Agora Coordinator.
package agorasdk

import (
	"encoding/json"
	"time"
)

// MessageType defines WebSocket message types for agent-coordinator
// communication. Values match the Python SDK MessageType enum.
type MessageType string

const (
	// Agent → Coordinator
	Register              MessageType = "REGISTER"
	Deregister            MessageType = "DEREGISTER"
	Speak                 MessageType = "SPEAK"
	RequestSpeak          MessageType = "REQUEST_SPEAK"
	Vote                  MessageType = "VOTE"
	Heartbeat             MessageType = "HEARTBEAT"
	TaskStatus            MessageType = "TASK_STATUS"
	TaskCompleted         MessageType = "TASK_COMPLETED"
	TaskFailed            MessageType = "TASK_FAILED"
	TaskStarted           MessageType = "TASK_STARTED"
	TaskProgress          MessageType = "TASK_PROGRESS"
	NewMotion             MessageType = "NEW_MOTION"
	RateLimitReport       MessageType = "RATE_LIMIT_REPORT"

	// Coordinator → Agent
	Welcome               MessageType = "WELCOME"
	SpeechAdded           MessageType = "SPEECH_ADDED"
	VoteConfirmed         MessageType = "VOTE_CONFIRMED"
	TaskAssigned          MessageType = "TASK_ASSIGNED"
	HeartbeatAck          MessageType = "HEARTBEAT_ACK"
	Error                 MessageType = "ERROR"
	DevilsAdvocateRequest MessageType = "DEVILS_ADVOCATE_REQUEST"
	DevilsAdvocateResponse MessageType = "DEVILS_ADVOCATE_RESPONSE"
	RateLimitWarning      MessageType = "RATE_LIMIT_WARNING"
	RateLimited           MessageType = "RATE_LIMITED"
	RateLimitReset        MessageType = "RATE_LIMIT_RESET"
	TaskBlocked           MessageType = "TASK_BLOCKED"
	TaskUnblocked         MessageType = "TASK_UNBLOCKED"
	TaskRetry             MessageType = "TASK_RETRY"
	GraphComplete         MessageType = "GRAPH_COMPLETE"
	GraphAborted          MessageType = "GRAPH_ABORTED"
)

// WSMessage is the generic WebSocket message envelope.
type WSMessage struct {
	Type      MessageType     `json:"type"`
	MotionID  *string         `json:"motion_id,omitempty"`
	AgentID   *string         `json:"agent_id,omitempty"`
	Timestamp time.Time       `json:"timestamp"`
	Payload   json.RawMessage `json:"payload,omitempty"`
}

// NewWSMessage creates a WSMessage with current UTC timestamp.
// If payload is non-nil, it is marshalled to JSON.
func NewWSMessage(msgType MessageType, payload interface{}) *WSMessage {
	m := &WSMessage{
		Type:      msgType,
		Timestamp: time.Now().UTC(),
	}
	if payload != nil {
		b, err := json.Marshal(payload)
		if err == nil {
			m.Payload = b
		}
	}
	return m
}

File: packages/agora-agent-sdk-rust/src/protocol.rs
//! Protocol types — WebSocket message types and WS envelope.
//!
//! MessageType variants match the Python SDK's `MessageType` enum.
//! WSMessage is the generic envelope used for all WS communication.

use serde::{Deserialize, Serialize};

/// WebSocket message type constants.
/// Agent → Coordinator variants are listed first, then Coordinator → Agent.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum MessageType {
    // Agent → Coordinator
    Register,
    Deregister,
    Speak,
    RequestSpeak,
    Vote,
    Heartbeat,
    TaskStatus,
    TaskCompleted,
    TaskFailed,
    TaskStarted,
    TaskProgress,
    NewMotion,
    RateLimitReport,
    // Coordinator → Agent
    Welcome,
    SpeechAdded,
    VoteConfirmed,
    TaskAssigned,
    HeartbeatAck,
    Error,
    DevilsAdvocateRequest,
    DevilsAdvocateResponse,
    RateLimitWarning,
    RateLimited,
    RateLimitReset,
    TaskBlocked,
    TaskUnblocked,
    TaskRetry,
    GraphComplete,
    GraphAborted,
}

/// Generic WebSocket message envelope.
/// Mirrors the Python SDK's `WSMessage` model.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct WSMessage {
    #[serde(rename = "type")]
    pub msg_type: MessageType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub motion_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payload: Option<serde_json::Value>,
}

impl WSMessage {
    /// Create a new WSMessage with the given type and optional payload.
    pub fn new(msg_type: MessageType, payload: Option<serde_json::Value>) -> Self {
        Self {
            msg_type,
            motion_id: None,
            agent_id: None,
            timestamp: Some(chrono::Utc::now().to_rfc3339()),
            payload,
        }
    }
}
